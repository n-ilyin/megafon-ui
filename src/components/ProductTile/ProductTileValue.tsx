import * as React from 'react';
import { cnCreate } from '../../utils/cn';
import './style/ProductTileValue.less';
import { IProductTileValue, IProductTileValueState } from './types';
import { productTileValueProps } from './propTypes';

const cn = cnCreate('mfui-product-tile-value');
class ProductTileValue extends React.PureComponent<IProductTileValue, IProductTileValueState> {
    static propTypes = productTileValueProps;

    constructor(props: IProductTileValue) {
        super(props);

        this.state = {
            prevValue: '',
            currentValue: props.value,
            isAnimating: false,
        };
    }

    static getDerivedStateFromProps(props: IProductTileValue, state: IProductTileValueState) {
        const isChangedValue = props.value !== state.currentValue;

        return {
            prevValue: isChangedValue ? state.currentValue : state.prevValue,
            currentValue: props.value,
            isAnimating: isChangedValue || state.isAnimating,
        };
    }

    componentDidUpdate() {
        const ANIMATION_DELAY = 50;

        if (typeof window === 'undefined') {
            return;
        }

        window.setTimeout(() => {
            this.setState({ isAnimating: false });
        }, ANIMATION_DELAY);
    }

    renderPrevPrice() {
        const { prevValue, isAnimating } = this.state;

        return (
            <span className={cn('prev', { hide: !isAnimating })}>
                {prevValue}
            </span>
        );
    }

    renderCurrentPrice() {
        const { currentValue, isAnimating } = this.state;

        return (
            <span className={cn('current', { hide: isAnimating })}>
                {currentValue}
            </span>
        );
    }

    render() {
        const { hAlign } = this.props;

        return (
            <span className={cn('', {'h-align': hAlign })}>
                {this.renderPrevPrice()}
                {this.renderCurrentPrice()}
            </span>
        );
    }
}

export default ProductTileValue;
