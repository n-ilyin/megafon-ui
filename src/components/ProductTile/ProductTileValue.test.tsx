import * as React from 'react';
import { shallow } from 'enzyme';
import ProductTileValue from './ProductTileValue';

const props = {
    value: '700 ₽',
    hAlign: 'center' as 'center',
};

describe('<ProductTileValue />', () => {
    it('renders ProductTileValue', () => {
        const wrapper = shallow(<ProductTileValue {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('renders ProductTileValue before animation', () => {
        const wrapper = shallow(<ProductTileValue {...props} />);

        wrapper.setProps({ value: '800 ₽'});
        expect(wrapper).toMatchSnapshot();
    });

    it('renders ProductTileValue after animation', () => {
        const wrapper = shallow(<ProductTileValue {...props} />);

        jest.useFakeTimers();
        wrapper.setProps({ value: '800 ₽'});
        jest.runOnlyPendingTimers();
        expect(wrapper).toMatchSnapshot();
    });
});
