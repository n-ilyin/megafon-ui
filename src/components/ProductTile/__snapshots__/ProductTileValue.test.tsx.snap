// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTileValue /> renders ProductTileValue 1`] = `
<span
  className="mfui-product-tile-value mfui-product-tile-value_h-align_center"
>
  <span
    className="mfui-product-tile-value__prev mfui-product-tile-value__prev_hide"
  />
  <span
    className="mfui-product-tile-value__current"
  >
    700 ₽
  </span>
</span>
`;

exports[`<ProductTileValue /> renders ProductTileValue after animation 1`] = `
<span
  className="mfui-product-tile-value mfui-product-tile-value_h-align_center"
>
  <span
    className="mfui-product-tile-value__prev mfui-product-tile-value__prev_hide"
  >
    700 ₽
  </span>
  <span
    className="mfui-product-tile-value__current"
  >
    800 ₽
  </span>
</span>
`;

exports[`<ProductTileValue /> renders ProductTileValue before animation 1`] = `
<span
  className="mfui-product-tile-value mfui-product-tile-value_h-align_center"
>
  <span
    className="mfui-product-tile-value__prev"
  >
    700 ₽
  </span>
  <span
    className="mfui-product-tile-value__current mfui-product-tile-value__current_hide"
  >
    800 ₽
  </span>
</span>
`;
